File: nhitomi/Interactivity/Triggers/ListTrigger.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Discord;

namespace nhitomi.Interactivity.Triggers
{
    public enum MoveDirection
    {
        Left,
        Right
    }

    public class ListTrigger : ReactionTrigger<ListTrigger.Action>
    {
        readonly MoveDirection _direction;

        public override string Name => $"Move {_direction}";

        public override IEmote Emote
        {
            get
            {
                switch (_direction)
                {
                    // left arrow
                    case MoveDirection.Left: return new Emoji("\u25c0");

                    // right arrow
                    case MoveDirection.Right: return new Emoji("\u25b6");
                }

                throw new ArgumentException(nameof(_direction));
            }
        }

        public ListTrigger(MoveDirection direction)
        {
            _direction = direction;
        }

        protected override void InitializeAction(Action action)
        {
            base.InitializeAction(action);

            action.Direction = _direction;
        }

        public class Action : ActionBase<IListMessage>
        {
            public MoveDirection Direction;

            readonly IServiceProvider _services;

            public Action(IServiceProvider services)
            {
                _services = services;
            }

            public override async Task<bool> RunAsync(CancellationToken cancellationToken = default)
            {
                if (!await base.RunAsync(cancellationToken))
                    return false;

                switch (Direction)
                {
                    case MoveDirection.Left:
                        Interactive.Position += 1;
                        break;

                    case MoveDirection.Right:
                        Interactive.Position += 1;
                        break;
                }

                return await Interactive.UpdateViewAsync(_services, Context, cancellationToken);
            }
        }
    }
}

File: nhitomi-discord/src/Triggers/ListTrigger.ts
import { ReactionTrigger } from "../Interactive/trigger";
import { InteractiveMessage } from "../Interactive/message";

function getEmoji(direction: "left" | "right") {
  switch (direction) {
    case "left":
      return "\u25c0";

    case "right":
      return "\u25b6";
  }
}

export type ListTriggerTarget = {
  position: number;
};

export class ListTrigger extends ReactionTrigger {
  constructor(readonly interactive: InteractiveMessage & ListTriggerTarget, readonly direction: "left" | "right") {
    super(interactive, getEmoji(direction));
  }

  protected async run() {
    switch (this.direction) {
      case "left":
        this.interactive.position--;
        break;

      case "right":
        this.interactive.position++;
        break;
    }

    return true;
  }
}
